### ISSUE TITLE
Cross-region ECR support

### ISSUE THREAD

**Issue by miguelfrde** [kubernetes/kubernetes#23298]
I’m having trouble to pull images from ECR on K8S 1.2. Before I was doing it with the imagePullSecrets field, but now I wanted to try the [new feature](http://kubernetes.io/docs/user-guide/images/#using-aws-ec2-container-registry) which sadly didn't work.

The policy for the minion IAM role I have looks just like the one on the [template](https://github.com/kubernetes/kubernetes/blob/master/cluster/aws/templates/iam/kubernetes-minion-policy.json) but it’s failing to pull the image:

```
$ kubectl describe pod myapp-api-celery-deployment-2551798148-bakj7
Name:		myapp-api-celery-deployment-2551798148-bakj7
Namespace:	default
Node:		ip-172-20-0-249.us-west-2.compute.internal/172.20.0.249
Start Time:	Mon, 21 Mar 2016 15:09:49 -0600
Labels:		app=myapp,component=api-celery,pod-template-hash=2551798148
Status:		Pending
IP:		10.244.1.3
Controllers:	ReplicaSet/myapp-api-celery-deployment-2551798148
Containers:
  myapp:
    Container ID:
    Image:		ACCOUNT.dkr.ecr.us-east-1.amazonaws.com/myapp:latest
    Image ID:
    Ports:		5672/TCP, 443/TCP, 80/TCP
    QoS Tier:
      cpu:	Burstable
      memory:	BestEffort
    Requests:
      cpu:		100m
    State:		Waiting
      Reason:		ImagePullBackOff
    Ready:		False
    Restart Count:	0
    Environment Variables:
      ENVIRONMENT:	aws
Conditions:
  Type		Status
  Ready 	False
Volumes:
  myapp-secret:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	myapp-secret
  default-token-0r10y:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	default-token-0r10y
Events:
  FirstSeen	LastSeen	Count	From							SubobjectPath		Type		Reason		Message
  ---------	--------	-----	----							-------------		--------	------		-------
  1m		1m		1	{default-scheduler }								Normal		Scheduled	Successfully assigned myapp-api-celery-deployment-2551798148-bakj7 to ip-172-20-0-249.us-west-2.compute.internal
  1m		35s		4	{kubelet ip-172-20-0-249.us-west-2.compute.internal}	spec.containers{myapp}	Normal		Pulling		pulling image "ACCOUNT.dkr.ecr.us-east-1.amazonaws.com/myapp:latest"
  1m		33s		4	{kubelet ip-172-20-0-249.us-west-2.compute.internal}	spec.containers{myapp}	Warning		Failed		Failed to pull image "ACCOUNT.dkr.ecr.us-east-1.amazonaws.com/myapp:latest": Error: Status 403 trying to pull repository myapp: "Your Authorization Token has expired. Please run 'aws ecr get-login' to fetch a new one."
  1m		33s		4	{kubelet ip-172-20-0-249.us-west-2.compute.internal}				Warning		FailedSync	Error syncing pod, skipping: failed to "StartContainer" for "myapp" with ErrImagePull: "Error: Status 403 trying to pull repository myapp: \"Your Authorization Token has expired. Please run 'aws ecr get-login' to fetch a new one.\""

  1m	7s	7	{kubelet ip-172-20-0-249.us-west-2.compute.internal}	spec.containers{myapp}	Normal	BackOff		Back-off pulling image "ACCOUNT.dkr.ecr.us-east-1.amazonaws.com/myapp:latest"
  1m	7s	7	{kubelet ip-172-20-0-249.us-west-2.compute.internal}				Warning	FailedSync	Error syncing pod, skipping: failed to "StartContainer" for "myapp" with ImagePullBackOff: "Back-off pulling image \"ACCOUNT.dkr.ecr.us-east-1.amazonaws.com/myapp:latest\""
```

Maybe it's worth noting the following:
1. The cluster was created with `kube-up.sh` but when it was created there was an existing `kubernetes-minion` role on AWS which didn't have that GetAuthorizationToken policy.
2. I added the policy after creating the cluster, but it didn't work.
3. I terminated the minion instances so that the autoscaler recreated them to see if maybe the policy was not refreshed for those instances. But that didn't work either.

**UPDATE**: as I mentioned on a comment below I didn't know ECR was available on `us-west-2` as well now. It seems that this works fine if both the ECR registry and the cluster are in the same region.

@justinsb

**Comment by druidsbane** [comment]
It seems like it currently only works on us-east-1.

**Comment by miguelfrde** [comment]
I just saw that ECR is now on `us-west-2` as well, so I tried putting my images there and it worked (since my cluster is on `us-west-2` as well. So it seems that for this to work both the cluster and the container registry have to be in the same region.

```
$ kubectl describe pod MYAPP-celery-api-deployment-653220251-5o4ym
Name:		MYAPP-celery-api-deployment-653220251-5o4ym
Namespace:	default
Node:		ip-172-20-0-249.us-west-2.compute.internal/172.20.0.249
Start Time:	Tue, 22 Mar 2016 10:15:52 -0600
Labels:		app=MYAPP,component=celery-api,pod-template-hash=653220251
Status:		Running
IP:		10.244.1.3
Controllers:	ReplicaSet/MYAPP-celery-api-deployment-653220251
Containers:
  MYAPP:
    Container ID:	docker://90a95009cf9e5045795f8efb00822a1f13abe602d9179c0fe9a636435216cf38
    Image:		ACCOUNT.dkr.ecr.us-west-2.amazonaws.com/MYAPP:latest
    Image ID:		docker://bde039654a9dbd63b991c2e8aa5cda8049bc87592116e66e67b5ada15528e8e0
    Ports:		5672/TCP, 443/TCP, 80/TCP
    QoS Tier:
      memory:	BestEffort
      cpu:	Burstable
    Requests:
      cpu:		100m
    State:		Running
      Started:		Tue, 22 Mar 2016 10:15:56 -0600
    Ready:		True
    Restart Count:	0
    Environment Variables:
      ENVIRONMENT:	aws
Conditions:
  Type		Status
  Ready 	True
Volumes:
  MYAPP-secret:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	MYAPP-secret
  default-token-0r10y:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	default-token-0r10y
Events:
  FirstSeen	LastSeen	Count	From							SubobjectPath		Type		Reason		Message
  ---------	--------	-----	----							-------------		--------	------		-------
  11s		11s		1	{default-scheduler }								Normal		Scheduled	Successfully assigned MYAPP-celery-api-deployment-653220251-5o4ym to ip-172-20-0-249.us-west-2.compute.internal
  10s		10s		1	{kubelet ip-172-20-0-249.us-west-2.compute.internal}	spec.containers{MYAPP}	Normal		Pulling		pulling image "ACCOUNT.dkr.ecr.us-west-2.amazonaws.com/MYAPP:latest"
  7s		7s		1	{kubelet ip-172-20-0-249.us-west-2.compute.internal}	spec.containers{MYAPP}	Normal		Pulled		Successfully pulled image "ACCOUNT.dkr.ecr.us-west-2.amazonaws.com/MYAPP:latest"
  7s		7s		1	{kubelet ip-172-20-0-249.us-west-2.compute.internal}	spec.containers{MYAPP}	Normal		Created		Created container with docker id 90a95009cf9e
  7s		7s		1	{kubelet ip-172-20-0-249.us-west-2.compute.internal}	spec.containers{MYAPP}	Normal		Started		Started container with docker id 90a95009cf9e
```

**Comment by justinsb** [comment]
OK, yes, that is a limitation for now ... same region only.

We also hit another bug.. if you had an existing IAM profile, kube-up doesn't update it.  This script will fix it:


```
#!/bin/bash

cat >rolepolicy <<EOF
{
        "Version": "2012-10-17", 
        "Statement": [
            {
                "Action": "s3:*", 
                "Resource": [
                    "arn:aws:s3:::kubernetes-*"
                ], 
                "Effect": "Allow"
            }, 
            {
                "Action": "ec2:Describe*", 
                "Resource": "*", 
                "Effect": "Allow"
            }, 
            {
                "Action": "ec2:AttachVolume", 
                "Resource": "*", 
                "Effect": "Allow"
            }, 
            {
                "Action": "ec2:DetachVolume", 
                "Resource": "*", 
                "Effect": "Allow"
            }, 
            {
                "Action": [
                    "ecr:GetAuthorizationToken", 
                    "ecr:BatchCheckLayerAvailability", 
                    "ecr:GetDownloadUrlForLayer", 
                    "ecr:GetRepositoryPolicy", 
                    "ecr:DescribeRepositories", 
                    "ecr:ListImages", 
                    "ecr:BatchGetImage"
                ], 
                "Resource": "*", 
                "Effect": "Allow"
            }
        ]
}
EOF

aws iam put-role-policy --role-name kubernetes-minion --policy-name kubernetes-minion --policy-document file://rolepolicy
```

**Comment by therc** [comment]
I want to take care of the cross-region issue in two steps. First, obtain credentials on demand, at pull time, rather than at startup. Second, register each region separately with the credential provider registry, because its API doesn't provide us with the actual URL needed for that call, so we can't have just one instance of our provider.

**Comment by therc** [comment]
Should I continue under this issue, with a different title, or create a new one?

**Comment by justinsb** [comment]
@therc you can use this issue or open a new one if you'd rather.  I think the IAM update issue is a separate issue.  I don't know if the "Your Authorization Token has expired." message is a bad message from ECR (i.e. == "you're using a token from another region"), or whether it indicates a problem with auto-refresh of tokens?

On fixing it, might it be possible to get the actual URL passed to your credentials provider?

**Comment by therc** [comment]
I'm assuming the message is Amazon's bug, but I'd have to verify. I can probably find a way to pass the URL through, but

1. that's an even more invasive change, so we'd need approval from @erictune 
1. it'd only be passed once every 12 hours, right? Still potentially useful for logging/debugging, not so much for auditing.

**Comment by rata** [comment]
On Thu, Mar 24, 2016 at 09:04:10AM -0700, Justin Santa Barbara wrote:
> @therc you can use this issue or open a new one if you'd rather.  I think the
> IAM update issue is a separate issue.  I don't know if the "Your Authorization
> Token has expired." message is a bad message from ECR (i.e. == "you're using a
> token from another region"), or whether it indicates a problem with
> auto-refresh of tokens?

I've used an ECR in the same region and leave the cluster idle for more than 12
hours, and after that I can use the ECR again. It worked just fine.

So, I'd guess that it's a bad message from ECR.

**Comment by mml** [comment]
@justinsb can you make the bug title more succint & correct?

**Comment by erictune** [comment]
@therc describe the invasive thing you are proposing to do?

**Comment by therc** [comment]
@erictune basically, what I suggested in https://github.com/kubernetes/kubernetes/issues/21783#issuecomment-192756850

> I'd like lazyDockerKeyring.Lookup() to stop calling Provide() blindly on all providers. Instead, it would use a new Locations(), or something similarly named, to grab just the list of URL patterns for registries the provider supports. Then, it would be the loop inside BasicDockerKeyring.Lookup() to call Provide(), but only after a match.

plus an additional change to Provide() so that it passes along the image URL. I'm not sure about @justinsb's exact use case here, because Provide() would be called only once every ~12 hours for ECR.

**Comment by erictune** [comment]
@therc so, each time I want to pull an image like `acct.dkr.ecr.$REGION.amazonaws.com/foo:bar`, the code will parse out $REGION, set up a session with $REGION, and get the credential?  

I agree that will be invasive, but it sounds necessary in order to support amazon's regional registry model.  So, fine by me.  

@dchen1107 any problem with making credential provider a bit more complicated?

**Comment by therc** [comment]
@erictune it will be only every 12 hours, for the first image in the region, but yeah, that's the idea. For AWS, we'd explicitly expand the regions and register each of them separately (we already keep a list of regions in the cloud provider). Or maybe register the wildcard location and populate the list of regions on demand, too, but we would still keep only one credential per region. For GCE, which has global credentials, we'd still keep the wildcard expressions for regions (*.gcr.io, etc.) and keep one credential for all of them.

**Comment by therc** [comment]
The keyring code turns out to be more complicated than it appeared at first. I changed it to keep a map of paths->providers instead of paths->credentials, in order to defer credential lookup, but the image puller's Pull() call in both Docker and rkt glue code builds a keyring on the fly by merging Docker credentials from the pod's secrets (if any) with the credentialprovider's keyring (by calling MakeDockerKeyring()).
I might fix that by creating a dummy provider to wrap the pod's secret credentials, but then a lot of the credentialprovider tests use the same pattern, too. Ugh.

**Comment by therc** [comment]
Can someone change the issue title to "Cross-region ECR support" and assign to me? This should definitely be in 1.3, if not 1.2.x.

**Comment by erictune** [comment]
Done

**Comment by jessesanford** [comment]
I am still having trouble getting things working in the same region. Is that still supported? See below (I have ensured that the policies have been created inside aws IAM control panel to allow the instance roles that the host nodes are booted with can access the ecr repositories)

```
$ kubectl describe pod rc-jenkins-master-a3e2w                                                                                           [2.2.4]
Name:				rc-jenkins-master-a3e2w
Namespace:			default
Image(s):			ACCOUNT_ID.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master:latest
Node:				ip-10-0-0-94.us-west-2.compute.internal/10.0.0.94
Labels:				env=devenv,name=jenkins,role=master
Status:				Pending
Reason:
Message:
IP:				10.2.89.2
Replication Controllers:	rc-jenkins-master (1/1 replicas created)
Containers:
  jenkins-master:
    Image:		ACCOUNT_ID.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master:latest
    State:		Waiting
      Reason:		PullImageError
    Ready:		False
    Restart Count:	0
Conditions:
  Type		Status
  Ready 	False
Events:
  FirstSeen				LastSeen			Count	From							SubobjectPath				Reason		Message
  Wed, 30 Mar 2016 02:20:35 -0400	Wed, 30 Mar 2016 02:20:35 -0400	1	{default-scheduler }										Scheduled	Successfully assigned rc-jenkins-master-a3e2w to ip-10-0-0-94.us-west-2.compute.internal
  Wed, 30 Mar 2016 02:20:35 -0400	Wed, 30 Mar 2016 02:20:35 -0400	1	{kubelet ip-10-0-0-94.us-west-2.compute.internal}	implicitly required container POD	Pulled		Container image "gcr.io/google_containers/pause:0.8.0" already present on machine
  Wed, 30 Mar 2016 02:20:35 -0400	Wed, 30 Mar 2016 02:20:35 -0400	1	{kubelet ip-10-0-0-94.us-west-2.compute.internal}	implicitly required container POD	Created		Created with docker id a221af90e7bd
  Wed, 30 Mar 2016 02:20:35 -0400	Wed, 30 Mar 2016 02:20:35 -0400	1	{kubelet ip-10-0-0-94.us-west-2.compute.internal}	implicitly required container POD	Started		Started with docker id a221af90e7bd
  Wed, 30 Mar 2016 02:20:35 -0400	Wed, 30 Mar 2016 02:20:59 -0400	4	{kubelet ip-10-0-0-94.us-west-2.compute.internal}	spec.containers{jenkins-master}		Pulling		Pulling image "ACCOUNT_ID.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master:latest"
  Wed, 30 Mar 2016 02:20:35 -0400	Wed, 30 Mar 2016 02:20:59 -0400	4	{kubelet ip-10-0-0-94.us-west-2.compute.internal}	spec.containers{jenkins-master}		Failed		Failed to pull image "ACCOUNT_ID.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master:latest": image pull failed for ACCOUNT_ID.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master:latest, this may be because there are no credentials on this request.  details: (API error (500): Get https://ACCOUNT_ID.dkr.ecr.us-west-2.amazonaws.com/v2/stn/jenkins-master/manifests/latest: no basic auth credentials
)
```

**Comment by jessesanford** [comment]
Wondering if the original impl was pulled out due to the refactoring going on and now the documentation is lagging?

**Comment by rata** [comment]
On Wed, Mar 30, 2016 at 07:04:53AM -0700, Jesse Sanford wrote:
> Wondering if the original impl was pulled out due to the refactoring going on and now the documentation is lagging?

No, it is working as it was reported here (if ECR registry and kubernetes
cluster are on the same region).

Make sure to give permissions on the ECR registry to the kubernetes-minion role
and run the script justinb posted.

**Comment by therc** [comment]
Does us-west-2 actually work with plain Docker?

**Comment by therc** [comment]
@jessesanford you can troubleshoot this in a couple of ways. Get us-west-2 credentials on your workstation. SSH into the host and run Docker manually with those. Does it work?

Since you're there, check kubelet logs. If your system uses journald, it might be as simple as `journalctl -t kubelet`. You might have to restart it with --v=2 or higher to increase verbosity.

**Comment by jessesanford** [comment]
@rata I am using the latest cloudformation template from the kube-aws project so the role in question for the kubelet (minion) node is: "Kubernetes-IAMRoleWorker-UNIQUEID" and in that role I have assigned the "AmazonEC2ContainerRegistryReadOnly" Managed policy that Amazon has available to all accounts. That policy contains the following policy document: 
```
{
	"Version": "2012-10-17",
	"Statement": [{
		"Effect": "Allow",
		"Action": [
			"ecr:GetAuthorizationToken",
			"ecr:BatchCheckLayerAvailability",
			"ecr:GetDownloadUrlForLayer",
			"ecr:GetRepositoryPolicy",
			"ecr:DescribeRepositories",
			"ecr:ListImages",
			"ecr:BatchGetImage"
		],
		"Resource": "*"
	}]
}
```

@therc the us-west-2 region is the region that @miguelfrde confirmed as working with this feature as of 8 days ago. I have tried k8s 1.2 and 1.3-alpha to no avail.

Is there any logging I can look for on the kubelet to find out if it is in fact trying to get the credentials from the amazon api? It's failing with a message leading me to believe that it's not recognizing that this is an ECR repository. If that's the case then it would not be trying to get the credentials correct?

**Comment by jessesanford** [comment]
Just to be clear. Running all other commands on this cluster is working just fine. I am able to launch publicly available images from the dockerhub without fail.

**Comment by rata** [comment]
On Wed, Mar 30, 2016 at 07:23:10AM -0700, Rudi C wrote:
> Does us-west-2 actually work with plain Docker?

It works for me (us-west-2a)

**Comment by jessesanford** [comment]
Here are the last few lines from the kubelet logs:

```
Mar 30 14:38:09 ip-10-0-0-94.us-west-2.compute.internal dockerd[1140]: time="2016-03-30T14:38:09.321588544Z" level=error msg="Handler for GET /images/932770329433.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master:latest/json returned error: No such image: 932770329433.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master:latest"
Mar 30 14:38:19 ip-10-0-0-94.us-west-2.compute.internal dockerd[1140]: time="2016-03-30T14:38:19.271257538Z" level=error msg="Handler for GET /images/932770329433.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master:latest/json returned error: No such image: 932770329433.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master:latest"
Mar 30 14:38:19 ip-10-0-0-94.us-west-2.compute.internal kubelet[957]: E0330 14:38:19.276369     957 kubelet.go:1455] Failed creating a mirror pod "kube-proxy-ip-10-0-0-94.us-west-2.compute.internal_kube-system": namespaces "kube-system" not found
Mar 30 14:38:19 ip-10-0-0-94.us-west-2.compute.internal dockerd[1140]: time="2016-03-30T14:38:19.273611339Z" level=error msg="Handler for GET /images/932770329433.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master:latest/json returned error: No such image: 932770329433.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master:latest"
Mar 30 14:38:19 ip-10-0-0-94.us-west-2.compute.internal dockerd[1140]: time="2016-03-30T14:38:19.296095582Z" level=error msg="Handler for POST /images/create returned error: Get https://932770329433.dkr.ecr.us-west-2.amazonaws.com/v2/stn/jenkins-master/manifests/latest: no basic auth credentials"
Mar 30 14:38:19 ip-10-0-0-94.us-west-2.compute.internal kubelet[957]: W0330 14:38:19.296316     957 manager.go:1897] Failed to pull image "932770329433.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master:latest" from pod "rc-jenkins-master-l5vjm_default" and container "jenkins-master": image pull failed for 932770329433.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master:latest, this may be because there are no credentials on this request.  details: (API error (500): Get https://932770329433.dkr.ecr.us-west-2.amazonaws.com/v2/stn/jenkins-master/manifests/latest: no basic auth credentials
Mar 30 14:38:19 ip-10-0-0-94.us-west-2.compute.internal kubelet[957]: )
Mar 30 14:38:19 ip-10-0-0-94.us-west-2.compute.internal dockerd[1140]: time="2016-03-30T14:38:19.299689866Z" level=error msg="Handler for GET /images/932770329433.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master:latest/json returned error: No such image: 932770329433.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master:latest"
```

**Comment by jessesanford** [comment]
I can confirm that I can retrieve the sts credentials from the amazon api for the role that is assigned to the worker node and then using those credentials get a login for the ecr repository and pull the image. So It is not a permissions issue on the amazon api side.

```
$  aws sts assume-role --role-arn "arn:aws:iam::ACCOUNT_ID:role/k8s1-3-IAMRoleWorker-13SGOW4FVLIP8" --role-session-name "MyRoleSessionName"
ASSUMEDROLEUSER arn:aws:sts::ACCOUNT_ID:assumed-role/k8s1-3-IAMRoleWorker-13SGOW4FVLIP8/MyRoleSessionName MASKED_CREDENTIALS_MATERIAL_HERE

$ export AWS_ACCESS_KEY_ID=MASKED
$ export AWS_SECRET_ACCESS_KEY=MASKED
$ export AWS_SESSION_TOKEN=MASKED

$ aws ecr get-login --region us-west-2
docker login -u AWS -p MASKED_CREDENTIALS_MATERIAL_HERE  -e none https://ACCOUNT_ID.dkr.ecr.us-west-2.amazonaws.com

$ docker pull ACCOUNT_ID.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master
Using default tag: latest
latest: Pulling from stn/jenkins-master
Digest: sha256:015cff0cf6e9ec1983066824bc1c8d29600a111b8fdc69f57d973b36415c7802
Status: Image is up to date for ACCOUNT_ID.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master:latest
```

**Comment by therc** [comment]
You should see this in the logs at startup, right before `Started kubelet`:

`plugins.go:56] Registering credential provider: aws-ecr-key`

Then later:

`provider.go:91] Refreshing cache for provider: *aws_credentials.ecrProvider`
`provider.go:91] Refreshing cache for provider: *credentialprovider.defaultDockerConfigProvider`

**Comment by jessesanford** [comment]
@therc I see:
`Mar 30 06:19:08 ip-10-0-0-94.us-west-2.compute.internal kubelet[957]: I0330 06:19:08.868983     957 plugins.go:56] Registering credential provider: .dockercfg`

but not:
`Registering credential provider: aws-ecr-key`

**Comment by therc** [comment]
Is your kubelet started with `--cloud-provider=aws`? Sounds like maybe it isn't. The ECR provider is not created automatically at startup, it is brought to life by the AWS cloud provider. Otherwise, it would cause long delays for non-AWS users.

**Comment by jessesanford** [comment]
@therc I assume it is running with the `--cloud-provider=aws` flag set because the systemd service references it:

```
core@ip-10-0-0-94 /tmp $ cat /etc/systemd/system/kubelet.service
[Service]
ExecStartPre=/usr/bin/mkdir -p /etc/kubernetes/manifests
ExecStart=/usr/bin/kubelet   --api_servers=https://10.0.0.50   --register-node=true   --allow-privileged=true   --config=/etc/kubernetes/manifests   --cluster_dns=10.3.0.10   --cluster_domain=cluster.local   --cloud-provider=aws   --kubeconfig=/etc/kubernetes/worker-kubeconfig.yaml   --tls-cert-file=/etc/kubernetes/ssl/worker.pem   --tls-private-key-file=/etc/kubernetes/ssl/worker-key.pem
Restart=always
RestartSec=10
[Install]
WantedBy=multi-user.target
```

**Comment by jessesanford** [comment]
@therc I also see other AWS specific features working like the automatic launching of ELB's for the service that I create that is based on a controller with a public image.

**Comment by jessesanford** [comment]
For the sake of completeness here is the kubelet startup logs:

```
Mar 30 15:54:56 ip-10-0-0-94.us-west-2.compute.internal systemd[1]: Stopping kubelet.service...
Mar 30 15:54:56 ip-10-0-0-94.us-west-2.compute.internal systemd[1]: kubelet.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Mar 30 15:54:56 ip-10-0-0-94.us-west-2.compute.internal systemd[1]: Stopped kubelet.service.
Mar 30 15:54:56 ip-10-0-0-94.us-west-2.compute.internal systemd[1]: kubelet.service: Unit entered failed state.
Mar 30 15:54:56 ip-10-0-0-94.us-west-2.compute.internal systemd[1]: kubelet.service: Failed with result 'exit-code'.
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal systemd[1]: Starting kubelet.service...
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal systemd[1]: Started kubelet.service.
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.191079   18175 aws.go:466] Zone not specified in configuration file; querying AWS metadata service
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.350333   18175 aws.go:575] AWS cloud filtering on tags: map[KubernetesCluster:kubernetes]
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.350493   18175 manager.go:128] cAdvisor running in container: "/system.slice/kubelet.service"
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.460853   18175 fs.go:108] Filesystem partitions: map[/dev/xvda9:{mountpoint:/ major:202 minor:9 fsType: blockSize:0}
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.486486   18175 manager.go:163] Machine: {NumCores:2 CpuFrequency:2800104 MemoryCapacity:3950592000 MachineID:261d69da
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.487190   18175 manager.go:169] Version: {KernelVersion:4.4.6-coreos ContainerOsVersion:CoreOS 1000.0.0 (MoreOS) Docke
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.489239   18175 server.go:798] Adding manifest file: /etc/kubernetes/manifests
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.489273   18175 server.go:808] Watching apiserver
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.656121   18175 plugins.go:56] Registering credential provider: .dockercfg
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.657207   18175 server.go:770] Started kubelet
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: E0330 15:55:04.657914   18175 kubelet.go:756] Image garbage collection failed: unable to find data for container /
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.662662   18175 server.go:72] Starting to listen on 0.0.0.0:10250
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.688808   18175 kubelet.go:777] Running in container "/kubelet"
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.691886   18175 factory.go:194] System is using systemd
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.803047   18175 factory.go:236] Registering Docker factory
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.803725   18175 factory.go:93] Registering Raw factory
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.916755   18175 manager.go:1006] Started watching for new ooms in manager
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.917803   18175 oomparser.go:183] oomparser using systemd
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.918393   18175 manager.go:250] Starting recovery of all containers
Mar 30 15:55:04 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:04.959775   18175 kubelet.go:885] Node ip-10-0-0-94.us-west-2.compute.internal was previously registered
Mar 30 15:55:05 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:05.021245   18175 manager.go:255] Recovery completed
Mar 30 15:55:05 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:05.096951   18175 manager.go:104] Starting to sync pod status with apiserver
Mar 30 15:55:05 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:05.096990   18175 kubelet.go:1953] Starting kubelet main sync loop.
Mar 30 15:55:05 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: I0330 15:55:05.118850   18175 provider.go:91] Refreshing cache for provider: *credentialprovider.defaultDockerConfigProvider
Mar 30 15:55:05 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: W0330 15:55:05.142656   18175 manager.go:1897] Failed to pull image "ACCOUNT_ID.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master
Mar 30 15:55:05 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: )
Mar 30 15:55:05 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: E0330 15:55:05.193088   18175 kubelet.go:1455] Failed creating a mirror pod "kube-proxy-ip-10-0-0-94.us-west-2.compute.internal_kube
Mar 30 15:55:15 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: W0330 15:55:15.136255   18175 manager.go:1897] Failed to pull image "ACCOUNT_ID.dkr.ecr.us-west-2.amazonaws.com/stn/jenkins-master
Mar 30 15:55:15 ip-10-0-0-94.us-west-2.compute.internal kubelet[18175]: )
```

**Comment by jessesanford** [comment]
And here it is with --v=4

```
Mar 30 16:14:12 ip-10-0-0-94.us-west-2.compute.internal systemd[1]: Started kubelet.service.
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.125676   19016 aws.go:575] AWS cloud filtering on tags: map[KubernetesCluster:kubernetes]
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.125729   19016 server.go:468] Successfully initialized cloud provider: "aws" from the config file: ""
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.125813   19016 manager.go:128] cAdvisor running in container: "/system.slice/kubelet.service"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.238006   19016 fs.go:108] Filesystem partitions: map[/dev/xvda9:{mountpoint:/ major:202 minor:9 fsType: blockSize:0}
/dev/xvda3:{mountpoint:/usr major:202 minor:3 fsType: blockSize:0} /dev/xvda6:{mountpoint:/usr/share/oem major:202 minor:6 fsType: blockSize:0}]
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.246237   19016 manager.go:163] Machine: {NumCores:2 CpuFrequency:2800104 MemoryCapacity:3950592000 MachineID:261d69da
c6af4b149f514c6168ac1a45 SystemUUID:EC254B21-6912-6AE1-A6C1-2B874B7454EF BootID:e7e7f47c-fbd4-40c8-97fd-95699e046e91 Filesystems:[{Device:/dev/xvda3 Capacity:1031946240} {Device:/dev/xvda6 Capacity:113229
824} {Device:/dev/xvda9 Capacity:28730269696}] DiskMap:map[202:16:{Name:xvdb Major:202 Minor:16 Size:16380329984 Scheduler:none} 202:0:{Name:xvda Major:202 Minor:0 Size:32212254720 Scheduler:none}] Networ
kDevices:[{Name:eth0 MacAddress:02:52:90:d1:8b:dd Speed:10000 Mtu:9001} {Name:flannel0 MacAddress: Speed:10 Mtu:8973}] Topology:[{Id:0 Memory:3950592000 Cores:[{Id:0 Threads:[0 1] Caches:[{Size:32768 Type
:Data Level:1} {Size:32768 Type:Instruction Level:1} {Size:262144 Type:Unified Level:2}]}] Caches:[{Size:26214400 Type:Unified Level:3}]}] CloudProvider:Unknown InstanceType:Unknown}
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.247179   19016 manager.go:169] Version: {KernelVersion:4.4.6-coreos ContainerOsVersion:CoreOS 1000.0.0 (MoreOS) Docke
rVersion:1.10.3 CadvisorVersion: CadvisorRevision:}
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.248501   19016 server.go:485] Using root directory: /var/lib/kubelet
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.254236   19016 server.go:718] cloud provider determined current node name to be ip-10-0-0-94.us-west-2.compute.intern
al
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.254307   19016 server.go:728] Sending events to api server.
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.254414   19016 server.go:798] Adding manifest file: /etc/kubernetes/manifests
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.254442   19016 file.go:47] Watching path "/etc/kubernetes/manifests"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.254461   19016 server.go:808] Watching apiserver
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.254641   19016 file.go:135] Reading config file "/etc/kubernetes/manifests/kube-proxy.yaml"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.264780   19016 config.go:252] Setting pods for source file
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.264868   19016 config.go:382] Receiving a new pod "kube-proxy-ip-10-0-0-94.us-west-2.compute.internal_kube-system"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.341479   19016 config.go:252] Setting pods for source api
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.341808   19016 config.go:382] Receiving a new pod "kube-proxy-ip-10-0-0-94.us-west-2.compute.internal_kube-system"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.341835   19016 config.go:382] Receiving a new pod "openvpn-mqw38_default"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.341850   19016 config.go:382] Receiving a new pod "rc-jenkins-master-f00wy_default"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.463781   19016 plugins.go:56] Registering credential provider: .dockercfg
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.464862   19016 plugins.go:262] Loaded volume plugin "kubernetes.io/aws-ebs"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.464882   19016 plugins.go:262] Loaded volume plugin "kubernetes.io/empty-dir"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.464893   19016 plugins.go:262] Loaded volume plugin "kubernetes.io/gce-pd"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.464904   19016 plugins.go:262] Loaded volume plugin "kubernetes.io/git-repo"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.464920   19016 plugins.go:262] Loaded volume plugin "kubernetes.io/host-path"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.464936   19016 plugins.go:262] Loaded volume plugin "kubernetes.io/nfs"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.464950   19016 plugins.go:262] Loaded volume plugin "kubernetes.io/secret"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.464969   19016 plugins.go:262] Loaded volume plugin "kubernetes.io/iscsi"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.464987   19016 plugins.go:262] Loaded volume plugin "kubernetes.io/glusterfs"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.465004   19016 plugins.go:262] Loaded volume plugin "kubernetes.io/persistent-claim"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.465018   19016 plugins.go:262] Loaded volume plugin "kubernetes.io/rbd"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.465033   19016 plugins.go:262] Loaded volume plugin "kubernetes.io/cinder"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.465047   19016 plugins.go:262] Loaded volume plugin "kubernetes.io/cephfs"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.465070   19016 plugins.go:262] Loaded volume plugin "kubernetes.io/downward-api"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.465086   19016 plugins.go:262] Loaded volume plugin "kubernetes.io/fc"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.465102   19016 plugins.go:262] Loaded volume plugin "kubernetes.io/flocker"
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.465152   19016 server.go:770] Started kubelet
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: E0330 16:14:13.466503   19016 kubelet.go:756] Image garbage collection failed: unable to find data for container /
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.473793   19016 server.go:72] Starting to listen on 0.0.0.0:10250
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.474029   19016 server.go:89] Starting to listen read-only on 0.0.0.0:10255
Mar 30 16:14:13 ip-10-0-0-94.us-west-2.compute.internal kubelet[19016]: I0330 16:14:13.474582   19016 server.go:726] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"ip-10-0-0-94.us-west-2.compute.internal", UID:"ip-10-0-0-94.us-west-2.compute.internal", APIVersion:"", ResourceVersion:"", FieldPath:""}): reason: 'Starting' Starting kubelet.
```

**Comment by therc** [comment]
Argh... programs that don't log their version as the very first thing when they come up. Which version of Kubelet is this?
The message should appear right after "AWS cloud filtering on tags: XYZ", as seen here:

https://github.com/kubernetes/kubernetes/commit/76e29ed455674361ca99e5c896d4499e1ba787ec

Actually, in your log, the "filtering" message was logged from line aws.go:575, but by the time my change landed, it was already at 593. Are you sure this is a recent binary? I suspect the message hasn't been logged from line 575 since this change in October: https://github.com/kubernetes/kubernetes/commit/9cd91d111d867f4a17d5796f028ba41f539c2747#diff-07ba008af9c76b0539556ff7fde3105e

**Comment by therc** [comment]
If this is CoreOS, try

```/usr/bin/kubelet --version=true
Kubernetes v1.1.2+1abf20d
```
That's from version 983.0.0. Ancient...

**Comment by jessesanford** [comment]
I was beginning to suspect the same thing:
```
/usr/bin/kubelet --version=true
Kubernetes v1.1.2+1abf20d
```

**Comment by jessesanford** [comment]
I think that this must be an issue with the kube-aws folks. It is installing 1.3 on the master controller and only 1.1.2 on the kubelet. I will dig farther and report back if I do find that they are the root cause. Thanks for your help and sorry for the misdirection.

**Comment by therc** [comment]
I don't set up my clusters kube-aws, but it seems like your system is using the kubelet binary that comes with CoreOS. Perhaps it notices it's there already and assumes you don't want another version?
Anyway, this was a good incentive for https://github.com/kubernetes/kubernetes/pull/23635

**Comment by erictune** [comment]
I updated the docs based on this thread.  Review welcome:
https://github.com/kubernetes/kubernetes.github.io/pull/286

**Comment by jessesanford** [comment]
Yep the issue was that kube-aws was old on my system so it was generating cloudformation templates that were referencing old cloud-config data which it uses to install kubelet etc. What's strange is that it was installing the correct version on the master controller. After updating to the latest version I can confirm that things are working as expected. Thank you.
